### PR TITLE
gmt: data-library_gui-refactor-wip branch PR

### PR DESCRIPTION
**Pull request by gmt**
As per your request (Nexela) here is a PR with my attempted refactoring of event/gui.

Let me know if you have any thoughts about it, I'm by no means attached to every detail here -- I figure it probably drags the brainstorm -> implementation bar somewhere near the 66% mark or so :)

### CODE REVIEW

**Comment by gmt** [comment]
This hunk in particular really could use some input from you... is there a reason for game.print() during event dispatch but log() during event setup?  I could imagine reasonable answers either way ("yes those are warnings and this is an error"; or "no, it just some sort of patchcruft/cut-paste slop that should be normalized") ... I think I'm not experienced enough with stdlib to make an educated guess here.

```-                    setmetatable(event, { __index = { _handler = handler } })
-                    local success, err = pcall(handler, event)
-                    if not success then
-                        game.print(err)
+                    local metahandler = {}
+                    setmetatable(event, {__index = metahandler})
+                    for _, handler in ipairs(handlers) do
+                        metahandler._handler = handler
+                        local success, err = pcall(handler, event)
+                        if not success then
+                            if game and game.print then
+                                game.print(err)
+                            else
+                                print(err)
+                            end
+                        end
                     end```

**Comment by gmt** [comment]
Looks like the test suite has some legitimate advice for me too :)

Is the test suite something I can run locally?  Make check says I need a "jq" whatever that is (but it just complains to stderr instead of failing)?

**Comment by Nexela** [comment]
You can run tests locally

Install busted
Make sure busted is on your path
In the main folder run `busted .`

To use the makefile it requires a few more things
Luacheck
Wget
Busted
jq

**Comment by Nexela** [comment]
If there is an error event library tries to print the error to the in game
console instead of crashing the game. However game is not always available
so we want to make sure the error is noticed

**Comment by gmt** [comment]
So, in other words game.print is sensible and you approve of checking game
(seemed like a pretty decent idea to me :))...

second question: failing game.print, do we log() or print()?  print()
always feels dirty, I don't even know what it really means (stderr? stdout?
factorio-current.log? undocumented? platform-dependent?).

Also, given that stdlib has its own logging facilities should I be using
those instead (throughout, I'd presume, if yes --- If so, is there a
best-practices-exemplar module somewhere I can ape?)

**Comment by gmt** [comment]
Thanks, for that (no doubt, there was some easy way for me to look this up; you telling me is definitely faster :P)

>  Nexela commented 21 minutes ago
> You can run tests locally
> 
> Install busted
> Make sure busted is on your path
> In the main folder run `busted .`
> 
> To use the makefile it requires a few more things
> Luacheck
> Wget
> Busted
> jq

**Comment by Nexela** [comment]
log()  writes to factorio current.log
print() writes to stdout (server console/terminal if started that way)

Logger don't use internally as it only writes to script-output

95% of the time log() is the best choice

**Comment by gmt** [comment]
Hm... well I copied that code (approximately) from the equivalent registry
manipulation code in Event.remove.... I'll think about this (for now I can
only trust-but-verify you, since I don't know a ton about lua internals).
Another stupid thing above is: if everything is bug-free, then there should
only ever be one i for which
Gui._registry[event_ids][gui_element_pattern][i] == handler.  So it should
really stop looping there, once we find / remove it (at that moment we also
have the answer to whether #reg is zero or not :))

But event registration isn't a super-performance-critical code-path,
obviously, and I'm concerned about the iterator equivalent to the problem
described in the insipid anecdote below....

The really coolest, most performance-elegance balanced idea would be: to
cook up an iterator that knows to either delete t[i] or increment i in
next(), and as soon as possible, return true iff we can also scrap the
pattern subtable.

This "count backwards for deletion" trick reminds me of the old-school
(like, almost before even my time :)) Algorithms 101 dirty-trick:

Professor finds fresh-faced first-year, teaches them the fundamentals of
Pascal, and asks "write a procedure which will delete multiple items from
an array-based list." on the Final. Inevitably the first year makes it like:

type:
Widget = <something>;
WidgetPointer = ^Widget;
WidgetList = array[1..100] of WidgetPointer;
WidgetRejector = function(w: Widget): boolean;

var:
w: WidgetList;
wr: WidgetRejector;

program:
procedure fetch_me_one_hundred_widgets(var w: WidgetList); forward;

procedure delete_widget_at(var w: WidgetList, idx: integer);
var i: integer;
begin
  for i := idx+1 to 100 do
    w[i-1] := w[i];
  end;
end;

procedure delete_rejected_widgets(var w:WidgetList, wr: WidgetRejector);
begin
 for i := 1 to 100
    if wr(w[i]) then
      delete_widget_at(w, i);
    end;
  end;
end;

begin
  fetch_me_one_hundred_widgets(w);
  delete_rejected_widgets(w, wr);
end.

Well, the professor says, great job, newbie!  You get a C.  Because, when
you call delete_widget_at(w, i), it puts the old value from w[i+1] into
w[i], and so on, so when you add one to i at the end of the loop in
delete_rejected_widgets, you are accidentally skipping the rejection test
for the next widget in the list (now at w[i]).

The point of all this being: there is a reason we don't program in Pascal
anymore... oh, no, wait, looks like we do again, my bad :)

File: stdlib/event/gui.lua [hunk 1]
```diff
@@ -2,6 +2,7 @@
 -- @module Gui
 -- @usage local Gui = require('stdlib/event/gui')
 
+require('stdlib/utils/table')
```
Comment (Nexela) [inline stdlib/event/gui.lua]: This is redundant as it is already required in core.

File: spec/setup/globals.lua [hunk 2]
```diff
@@ -1,4 +1,4 @@
-_G.serpent = require("serpent")
+_G.serpent = require("stdlib/utils/serpent")
```
Comment (Nexela) [inline spec/setup/globals.lua]: While I do like this change, this is a note to self to remove serpent from utils and put it in spec/setup as factorio already includes a custom version of serpent.

File: stdlib/event/gui.lua [hunk 3]
```diff
@@ -2,42 +2,15 @@
 -- @module Gui
 -- @usage local Gui = require('stdlib/event/gui')
 
-require('stdlib/event/event')
+require('stdlib/utils/table')
```
Comment (Nexela) [inline stdlib/event/gui.lua]: Redundant call as it is already included in core.

File: stdlib/event/gui.lua [hunk 4]
```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
...
-                        game.print(err)
+                    local metahandler = {}
+                    setmetatable(event, {__index = metahandler})
+                    for _, handler in ipairs(handlers) do
+                        metahandler._handler = handler
+                        local success, err = pcall(handler, event)
+                        if not success then
+                            if game and game.print then
+                                game.print(err)
+                            elseif log then
```
Comment (Nexela) [inline stdlib/event/gui.lua]: log will always be available, it is a protected function

File: stdlib/event/gui.lua [hunk 5]
```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
...
+        return Gui
+    end
+
+    local reg = Gui._registry[event_ids][gui_element_pattern]
+    for i = #reg, 1, -1 do
+        if reg[i] == handler then
+            table.remove(reg, i)
         end
     end
+    if not table.any(reg, function (v) return true; end) then -- luacheck: ignore v
```
Comment (Nexela) [inline stdlib/event/gui.lua]: `if table.size(reg) == 0 then` is still a much faster check, PS  for the future function() instead of function(v) since you are not using any parameters

File: stdlib/event/gui.lua [hunk 6]
```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
...
+        Gui._registry[event_ids][gui_element_pattern] = nil
+    end
+    if not table.any(Gui._registry[event_ids], function(v) return true; end) then -- luacheck: ignore v
```
Comment (Nexela) [inline stdlib/event/gui.lua]: Again here,  `if table.size(Gui._registry[event_ids]) == 0 then`

File: stdlib/event/gui.lua [hunk 7]
```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
...
+-- @tparam defines.events event_ids valid values are `defines.events.on_gui_*` from @{defines.events},
+-- or a list of them.
+-- @tparam string gui_element_pattern the name or string regular expression to match the gui element
+-- @tparam function handler the function to call when the event is triggered
 -- @return (<span class="types">@{Gui}</span>)
-function Gui.remove(event_id, gui_element_pattern)
-    fail_if_missing(event_id, "missing event argument")
+function Gui.register(event_ids, gui_element_pattern, handler)
+    fail_if_missing(event_ids, "missing event name argument")
+    fail_if_missing(gui_element_pattern, "missing gui name or pattern argument")
```
Comment (Nexela) [inline stdlib/event/gui.lua]: ` fail_if_missing(Is.String(gui_element_pattern), "missing gui pattern string")`

File: stdlib/event/gui.lua [hunk 8]
```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
...
+
+    if type(event_ids) == "table" then
+        for _, v in pairs(event_ids) do
+            Gui.register(v, gui_element_pattern, handler)
+        end
+        return Gui
+    end
 
     if type(gui_element_pattern) ~= "string" then
```
Comment (Nexela) [inline stdlib/event/gui.lua]: This check should go above the recursive loop because,    fail first :)   however if the above fail_if_missing change is added this check can go the way of the Dodo bird

File: stdlib/event/gui.lua [hunk 9]
```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
...
+    fail_if_missing(event_ids, "missing event argument")
+
+    if type(event_ids) == "table" then
+        for _, v in pairs(event_ids) do
+            Gui.remove(v, gui_element_pattern, handler)
+        end
+        return Gui
+    end
+
+    if gui_element_pattern and type(gui_element_pattern) ~= "string" then
```
Comment (Nexela) [inline stdlib/event/gui.lua]: `if Is.String(gui_element_pattern) then`

File: stdlib/event/gui.lua (hunk 3 above)
Comment (gmt) [reply stdlib/event/gui.lua]: I am pretty confused about how core actually works.  I don't know:

* Is there some meta-magic or convention that's supposed to guarantee core stuff is in the module's _ENV (that is, can/should certain things in core be treated as globals in module code)?
* Does the core metatable provide a certain API to consumers of stdlib modules?  If so, what?
* What is the correct/best way to build a new stdlib module (including the core stuff)?
* What is the "friend" API surface of core -- put another way, if I am coding a module, are there things in core I should /not/ be consuming as they are considered implementation details that might change?
* Is there official documentation for core?

That's a lot of questions so feel free to cherry-pick :)  Any insights would be appreciated.

File: stdlib/event/gui.lua (hunk 4 above)
Comment (gmt) [reply stdlib/event/gui.lua]: ACK, deparanoiafied downstream.

File: stdlib/event/gui.lua (hunk 5 above)
Comment (gmt) [reply stdlib/event/gui.lua]: >>`table.any(reg, function(v) return true; end)`
>
>`if table.size(reg) == 0 then` is still a much faster check

I am a bit mixed up about table internals, different versions of lua, etc...  For now I'm just leaving it but I do intend to reevaluate after my next visit to the clue depot.  Maybe we should add something like this to stdlib/event/table?:
```
function table.empty(t)
  return next(t)==nil
end
```

> function() instead of function(v) since you are not using any parameters

Lua will just let a function's consumer throw arbitrary parameters into a black hole like that?  *shudder*... Would luacheck be happy about it?

File: stdlib/event/gui.lua (hunk 7 above)
Comment (gmt) [reply stdlib/event/gui.lua]: ACK, brilliant, pushed downstream.

File: stdlib/event/gui.lua (hunk 8 above)
Comment (gmt) [reply stdlib/event/gui.lua]: ACK, also downstream.

File: stdlib/event/gui.lua (hunk 9 above)
Comment (gmt) [reply stdlib/event/gui.lua]: Fixed downstream... Now several places that incorrectly accept Is.False garbage inputs are starting to look more like bugs and less like "when in rome" situations to me, but that could turn into a pretty big audit so I'm ignoring it for now.

File: stdlib/event/gui.lua [hunk 10]
```diff
@@ -99,6 +99,9 @@ end
 -- @return (<span class="types">@{Gui}</span>)
 function Gui.remove(event_ids, gui_element_pattern, handler)
     fail_if_missing(event_ids, "missing event argument")
+    if not (Is.Nil(gui_element_pattern) or Is.String(gui_element_pattern)) then
```
Comment (Nexela) [inline stdlib/event/gui.lua]: It is always mostly my fault
```lua
fail_if_missing(Is.Nil(gui_element_pattern) or Is.String(gui_element_pattern), guil_element_pattern argument must be a string when present)
````
This will have consistent stack trace
Ps think of fail_if_missing as just fail_if_falsy(condition, message),  I may change it at some point

Comment (gmt) [reply stdlib/event/gui.lua]: Understood, sorted downstream.

File: stdlib/event/gui.lua (hunk 5 above)
Comment (Nexela) [reply stdlib/event/gui.lua]: Well it isn't really a black hole  function(v) creates a local variable v and puts whatever the function passes into it,  if you are not using v at all no point even adding it to the parameters just to ignore it. and luacheck will be happy.

table_size is a factorio function which returns a c++ count of keys which is faster than calling next(t), table.size is just a re-mapping to add it to the table mutates, stdlib also accounts for the potential lack of table_size (ie for busted tests) by treating table.size as table.count_keys if table_size is nil.

If it is not already present I will add table.isempty and Is.Empty, as well as table.clear

Comment (gmt) [reply stdlib/event/gui.lua]: Nexela said:
> table_size is a factorio function which returns a c++ count of keys which is faster than calling next(t), table.size is just a re-mapping to add it to the table mutates, stdlib also accounts for the potential lack of table_size (ie for busted tests) by treating table.size as table.count_keys if table_size is nil.

That explains everything.  I thought we were talking about a lua intrinsic, but I must have been thinking of getn.

Now that I think about it, in working on this stuff, I may have discovered that next is broken for certain factorio userdata things.  For example, I am pretty sure that next(game.players, next(game.players)) == "__self" or something crazy like that.  My conclusion was: if x,y,z = pairs(game.players), somehow, x ~= next.   I could be mixed up; regardless, I think something tricky is happening when you iterate over game.players... anyhow I digress.

Comment (gmt) [reply stdlib/event/gui.lua]: > If it is not already present I will add table.isempty and Is.Empty, as well as table.clear

That would be fantastic, thanks/tia (depending on whether or not you already did so, respectively :))

File: stdlib/event/gui.lua (hunk 3 above)
Comment (Nexela) [reply stdlib/event/gui.lua]: Core is just the core, All the stdlib classes have an __index into core using setmetatable

When core is first required it requires some files
table/string/Iter/Is (etc) and makes them global (in the case of table and string it mutates the existing global)
core.lua
```lua
local Core = {}
Core.func = function() return true end
return Core
```
data.lua
```lua
local Data = setmetatable({}, {__index = require("core")}) --require checks in package.loaded for Core if it exists it returns that value (the return return from core.lua is the value), otherwise require  the file and stick the return here
return Data
```
file.lua
local Data = require("data")
if Data.func() then print("good times") end
```

Comment (gmt) [reply stdlib/event/gui.lua]: Nexela said:
> data.lua:
> ```lua
> local Data = setmetatable({}, {__index = require("core")}) -- ...
> return Data
> ```

I guess what is confusing me is, what is the purpose of using __index here?

For example, doesn't this mean that after a factorio mod does  `Data = require('stdlib/data/data')`, then `data._module_name == "Core"`, `data.VALID_FILTER(blah) == blah.valid`, etc?  I guess that is a secondary side effect, that we are assuming won't be messed around with by stdlib consumers?

But, then, what are all those _x virtuals actually doing for the stdlib module that inherits them?  Or is that too confusing to be explainable in this context, i.e., I should just read more stdlib code if I want to grok this?

File: stdlib/event/gui.lua (hunk 5 above)
Comment (gmt) [reply stdlib/event/gui.lua]: I wrote:
> Now that I think about it, in working on this stuff, I may have discovered that next is broken for certain factorio userdata things. For example, I am pretty sure that next(game.players, next(game.players)) == "__self" or something crazy like that. My conclusion was: if x,y,z = pairs(game.players), somehow, x ~= next. I could be mixed up; regardless, I think something tricky is happening when you iterate over game.players... anyhow I digress.

But, I looked again, and, in a single-player game,

serpent.line({next(game.players)}) => {"__self", "userdata..."}
serpent.line({next(game.players, next(game.players))}) => {}

So, it seems, game.players is a completely normal table containing an array of player userdata objects, precisely as I would expect it to be.  Instead the real problem appears to be a pebcak, specifically, my tendency, (amusingly identifiable as a bug in the code quoted above) to sometimes forget that next is returning two values rather than a two-element array.

**Comment by Nexela** [comment]
I am pretty sure some of your comments/questions got swallowed in the litany of changes,   If there are any unanswered/unresolved post them under this one and I will do my best to answer

Most breaking change in the lastest data-library branch,  Is.Stuff has been moved from global name space to Core.Is   (or can be required locally) local Is = require("stdlib/utils/is") and serpent has been moved to spec/world/utils    Since it is not needed for factorio and could potentially cause issues due to custom serpent handling.
Also some other things have been moved from Core to globals.lua.  .luacheckrc has been updated to compensate

**Comment by gmt** [comment]
Nexela said:
> I am pretty sure some of your comments/questions got swallowed in the litany of changes, If there are any unanswered/unresolved post them under this one and I will do my best to answer

10-4, I think likewise some of yours.  I am still cooking up a reply to the most important one, however, in my opinion.

Nexela wrote:
> ...Is.Stuff has been moved...
> [and various other stuff about various possibly breaking changes upstream]

No problem, that is what git and the test suite are for.  I pretty-much eat merge conflicts for breakfast.

**Comment by gmt** [comment]
Nexela asked:
> What are some of your complaints about the current version of Event and Gui?

Disclaimer: Below, I've tried to answer this clearly and concisely.  The following is hopefully clear; for the concise version, please read only the first sentence of every paragraph :P

I have two:
* First: it turns Event._registry into a confusing muddle.

  Imagine if we were coding this in C instead of lua.  We'd have something like:

  ```C
  typedef void (*single_event_handler)(event_info e);
  typedef single_event_handler multiple_event_handlers[];
  typedef struct registrant {
    event_id eid;
    union {
      struct {
        multiple_event_handlers *handlers;
      } event_module_handler_stuff;
      struct {
        char *filter;_stuff
        single_event_handler handler;
      } gui_module_handler_stuff;
    }
  };
  struct registrant _registry[] = ...
  ```

  Putting this nicely, once we've built that data structure, in any language, I think we should really be taking a hard look at how we've structured our code and whether there might be a cleaner way to modularize our data.

* Second: the "friend" relationship between the event and gui modules prevents stdlib from converting these modules from my-god-its-full-of-globals into lua5.2+-standard modules, emitting a single lua multi-value.

Ultimately, my patch assumes both issues share a common cause: the sharing of the registry.  Frankly, it looks like the event module was sitting around minding its own business, and gui came along and mugged it, stealing it's private registry and jamming its own incompatible data structures in sideways.  Sure, they both broadcast events to registered listeners with identical signatures, but that is where the similarities end, upstream.

If the registries are decoupled, we can more-or-less trivially make both gui and event private modules, each with lexically private registry variables.  (grep my branch for "_registry" & you'll see what I mean -- the only remaining consumers internally are the test suite (which would need rewriting in order to finish the job)).

As for desyncs: I know less about that, but I can't see how it could make any difference.  Either way, there is only one logical registry singleton per lua instance, per module.  Any other behavior is clearly a bug.  I think decoupling registries makes any such bug harder to create and easier to fix.

To summarize "exactly that the heck this does" here is how events flow upstream: (with "->>" indicating multiplexing):

```
  [factorio] <-+-> event <->> mod code
               |     |
               |     v
               +-> gui <->> mod code
```

No, that's not a mistake, there are two registration pathways depending, iiuc, on load order.  I still don't fully understand it, to be honest -- at some point I just threw up my hands started working on fixing it instead :)

And here is how they flow in my branch:

```
  [factorio] <-> event <-+->> mod code
                         |
                         +->> gui <->> mod code
```

So basically event becomes the sole registrant with factorio, and gui becomes
a consumer just like a mod.  gui then offers an extra demultiplexing service
to consumers who want to filter the gui events.  I see it kind of like
daisy-chaining two IP routers.  Each router looks at the message and decides who
are the list of recipients that should see it, and then either eats it or forwards it along as appropriate.

I do have one misgiving about my approach: in retrospect, I am starting to feel that the same result could have been arrived at more elegantly by moving the filtering feature from the gui module up into the event module, and reducing the gui module into a kind of syntactic sugar.  Anyhow, here, I'm thinking the perfect is maybe the enemy of the good.  Stdlib can always revisit this later & this moves things in the right direction.

**Comment by gmt** [comment]
> I will probably take a shot at this pretty soon either way.

First I think I should try to get to a test branch with fully decoupled, pure-local event and gui module using the hierarchical muxing.  Not sure how much freedom there is to break things with all the global/local stuff going on but I'm get the impression you're working on a depreciation scheme as I speak...

**Comment by Nexela** [comment]
> I get the impression you're working on a depreciation scheme as I speak...

If you mean,   you have error make it local fool :) then yes

In the current data-library branch Everything but event and gui is mandatory local  (with the exception of a few functions in utils/globals.lua and the string/table/math mutates

The registry (Event._registry) itself HAS to be global to be shared amongst scripts there is no way around it, 

file_a
```
Event.register(1, funca)
```
file_b
Event.register(1, funcb)
```
If the registry is local File_b will see an empty _registry and run script.on_event(1, Event.Dispatch) which will only contain the _registry for funcb so funca never runs (table_size(Event.registry) == 1)
The opposite will happen with remove! it will nil out the script.on_event and BAM no more Events are run from other scripts

However Event._registry can be de-coupled from Event and put in the _G namespace and referenced that way

in event.lua
```
_G._event_registry = _G._event_registry or {}
local Event = {}
Event._registry = _G._event_registry
```
in Gui
_G._event_registry = _G._event.registry or {}
local Gui = {}
Gui._registry = _G._event_registry
```

In summary These are the implementation changes I would accept for Event/Gui
Detach the _registry from Event and Gui and make it global (only Event and Gui currently will ever need access to it)
use the same detached registry for STORING the events
local Gui and Event
(And if wanted Incorporate Gui into Event or even Merge into existing funcitons for less duplicity, and use Gui for syntactic sugah!  Gui.on_click() being shorthand for Event.register(gui_event, function(handler))

**Comment by gmt** [comment]
> in Gui
> 
> _G._event_registry = _G._event.registry or {}
> local Gui = {}
> Gui._registry = _G._event_registry
> In summary These are the implementation changes I would accept for Event/Gui
> Detach the _registry from Event and Gui and make it global (only Event and Gui currently will ever need access to it)
> use the same detached registry for STORING the events
> local Gui and Event
> (And if wanted Incorporate Gui into Event or even Merge into existing funcitons for less duplicity, and use Gui for syntactic sugah! Gui.on_click() being shorthand for Event.register(gui_event, function(handler))

Hm... well I don't mean to be stubborn but I still suspect that when everything shakes out you'll be converted to my opinion that there's no need for it to be global or even accessible outside of event.lua.  But that's an implementation detail that will follow naturally from the overall structure, which is still kind of up in the air.  I'll keep hacking on this branch and we'll see what shakes out.

For now, I think the whole branch is overdue for a rebase and overhaul.

Since you prefer a single registry, clearly I should be shooting for a "syntactic-sugar" arrangement in favor of hierarchical-dispatch.  The more I think about that, the more I tend to agree with you:  Why should gui events be singled out for this pattern-matching ability, when, with some fairly trivial cut-and-paste development, we could empower any consumer of any event to do the same to any property that has a string representation?  Maybe we could do even better.... but there be bike sheds.

I'll keep at it.

**Comment by Nexela** [comment]
Ok I had some time to think about the current implementation how it is handled and having a separate handler registry for Gui would be ok since the actual gui event would still be in the Events registry .
_registery = {
tick = Event.dispatch() -- which uses Event._registry
on_click = Gui.dispatch() -- which uses Gui._registry
}
Which also might point to a potential fatal flaw I will have to look into:  (PS I have not spent much time studying Gui)
```lua
Gui.register(on_click, funca)
Event.register(on_click, funcb)
```

Ok I guess you are right about the registries not needing to be global.... damn table references
`local E = require("event")   -- package.loaded.event = Event E = address of package.loaded.event`

So I guess issues would only arise of someone feels like clearing package.loaded.event

**Comment by gmt** [comment]
Closing in anticipation of new event-enhancements-based PR
